### PR TITLE
Update NWjs to 0.67.1 and Node to 16.17.0 LTS

### PR DESCRIPTION
**Pull request by haslinghuis**
https://nwjs.io/blog/v0.67.1/
https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V16.md#16.17.0

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

File: src/css/dark-theme.css
Comment (McGiverGim) [inline src/css/dark-theme.css:524]: It will be good if you explain in the PR why you remove this. I suppose the graphic glitches you are seeing come from this?

Comment (haslinghuis) [reply src/css/dark-theme.css:524]: Yes the glitch is removed by removing these lines as they are available here:

https://github.com/betaflight/betaflight-configurator/blob/master/src/css/tabs/osd.css#L486-L508

Comment (McGiverGim) [reply src/css/dark-theme.css:524]: There is no lnear-gradient in this code. Maybe it was a way to make the preview "darker" in the dark theme? Just after that, you have the same but only when active, maybe this was the reason of the square when drag&drop?

Comment (haslinghuis) [reply src/css/dark-theme.css]: Just checked again on Ubuntu 20.04 LTS and 22.04 LTS. This part is somehow not an issue anymore so removed now.
The fix in #2913 is still needed to remove the square on drag&drop.

**Review by McGiverGim — APPROVED** [review]

**Comment by meoso** [comment]
nwjs `0.63.x`, `0.64.x`, and `0.65.0` are producing faulty file rights from the packages.  (i tested .deb).

`0.62.2` is the last one to work.

The result is 
image

due to `group` and `others` rights are set `none`:
image

instead of the expected `read-only`:
image

**Comment by haslinghuis** [comment]
@meoso how about current version?

**Comment by haslinghuis** [comment]
@meoso Thanks. Let's wait for a newer version instead.

https://nwjs.io/blog/

**Comment by McGiverGim** [comment]
@haslinghuis I'm not too sure if we are a little off or I'm confusing some versions...

We use the `.nvrmc` file to define the Node version:
https://github.com/betaflight/betaflight-configurator/blob/38688be1f817f80f952c5233625612c6dc550cf6/.nvmrc#L1
https://github.com/betaflight/betaflight-configurator/blob/38688be1f817f80f952c5233625612c6dc550cf6/package.json#L50-L52

But we ask for the 0.62 Node.js while building:
https://github.com/betaflight/betaflight-configurator/blob/38688be1f817f80f952c5233625612c6dc550cf6/gulpfile.js#L50-L51

According to the log, since 0.59.1 the Node.js version is v17.3.0: https://nwjs.io/blog/v0.59.1/

So, maybe I'm confusing something, but I think we are not using the correct vesion? Or the backwards compatibility is assured?

**Comment by haslinghuis** [comment]
Agree we should solve it in tandem and should add Node.js v17.3.0 or perhaps even v18.3.0 if we add NWjs 0.65,1

See: https://github.com/nwjs/nw.js/blob/nw65/CHANGELOG.md

**Comment by haslinghuis** [comment]
@meoso please check?

**Comment by meoso** [comment]
* checked out `32e8edc`
* `nvm install 18.3.0; nvm use 18.3.0; rm -rf node_modules ; yarn install`
* built release; installed release
* user/executable rights do not apply correctly still for `group` & `other`
image

**Comment by haslinghuis** [comment]
Thanks @meoso for checking.

Bonkers. Azure ubuntu-2204 uses Node 16.15.1 so even it runs locally we can't have (release) builds to match NW.js with Node version.

https://github.com/actions/virtual-environments

So we'll have to postpone

**Comment by haslinghuis** [comment]
Thanks @meoso for checking.

@McGiverGim Azure ubuntu-2204 uses Node 16.15.1 so even it runs locally we can't have (release) builds to match now working NW.js version with the suggested Node version.

https://github.com/actions/virtual-environments

**Comment by McGiverGim** [comment]
Maybe we can try to use this? https://github.com/actions/setup-node

**Comment by meoso** [comment]
* success with `711ea41`!  
* installs proper and assigns correct rights
* executes proper (did not test functionality)
image

**Comment by McGiverGim** [comment]
@haslinghuis I'm not too sure, but I suspect the problem is only with Mac. It fails, and the others builds are cancelled (except the Android one that is faster). Do you have a way to test the Mac build in local?

**Comment by haslinghuis** [comment]
@McGiverGim might need osx to add it:
```
[5/5] Building fresh packages...
Done in 21.99s.
mark@pc:~/dev/pr/update_nwjs/betaflight-configurator$ yarn add appdmg
yarn add v1.22.17
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
error macos-alias@0.2.11: The platform "linux" is incompatible with this module.
error Found incompatible module.
```

**Comment by McGiverGim** [comment]
No really... you can change it manually in the package.json file: https://github.com/betaflight/betaflight-configurator/pull/2956

But I don't see changes in the different dependencies, so I think it will not work neither...

**Comment by meoso** [comment]
this is how i did it elsewhere
```
  "config": {
    "platformDependentModules": {
      "darwin": [
        "appdmg@^0.6.2"
      ]
    }
  }
```

**Comment by meoso** [comment]
> Do you want to test this code? Here you have an automated build:  [Betaflight-Configurator-Linux](https://github.com/betaflight/betaflight-configurator/suites/7067418619/artifacts/278830348)

image :+1:

**Review by blckmn — APPROVED** [review]

**Review by McGiverGim — DISMISSED** [review]

**Review by blckmn — APPROVED** [review]

**Comment by haslinghuis** [comment]
@meoso did you also check nwjs 0.66.0 on configurator?

**Comment by meoso** [comment]
> Do you want to test this code? Here you have an automated build:  [Betaflight-Configurator-Linux](https://github.com/betaflight/betaflight-configurator/suites/7427728403/artifacts/303370916) 

Debian 11: 👍
image

**Comment by haslinghuis** [comment]
@meoso updated NWjs.

**Comment by meoso** [comment]
.deb fails install on Debian 11 due to unsupported compression
image

portable works

**Comment by meoso** [comment]
> Do you want to test this code? Here you have an automated build: [Betaflight-Configurator-Linux](https://github.com/betaflight/betaflight-configurator/suites/7448074693/artifacts/304692732) 

same problem, no change. 
```
$ sudo dpkg -i betaflight-configurator_10.9.0-debug-0c21594_amd64.deb 
dpkg-deb: error: archive 'betaflight-configurator_10.9.0-debug-0c21594_amd64.deb' uses unknown compression for member 'control.tar.zst', giving up
```

**Comment by haslinghuis** [comment]
@meoso Seems Debian is somewhat compelled adding `zst` support for `dpkg` 

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=892664

**Comment by meoso** [comment]
Seems so, but Debian being "stable" is always extremely slow to adopt as you can see talks started in 2018 and yet to be released. I suspect no fixes for many months or more. :(

**Comment by meoso** [comment]
i hate to say it, but even when i `yarn gulp clean ; yarn gulp release --linux64` locally, the file rights reverted to prior failures as well. :(
commit `56216b8c`.

node/nwjs, always a mess! :(

image

**Comment by McGiverGim** [comment]
I don't understand the problem exactly... some file in the new version uses some format that when decompressing is giving the error?
The problem is in the creation of the deb file or in the contents of it?

**Comment by meoso** [comment]
> I don't understand the problem exactly... some file in the new version uses some format that when decompressing is giving the error? The problem is in the creation of the deb file or in the contents of it?

The compression method used to compress the .deb file.  
Old standard: `Debian binary package (format 2.0), with control.tar.xz, data compression xz`
Method in this PR: `Debian binary package (format 2.0), with control.tar.zs, data compression zst` which is not supported by Debian (or at least by Debian Stable 11)

**Comment by haslinghuis** [comment]
@McGiverGim

Ubuntu has introduced `zst` compression in `dpkg` command. Debian does not have support for this protocol.

**Comment by McGiverGim** [comment]
Then the problem is with the new version of Ubuntu used to build the deb file, not with this PR itself, right?

Then the only way to fix it is in this code:
https://github.com/betaflight/betaflight-configurator/blob/1c775439a1539e93dd543535621b9b06c362bf8b/gulpfile.js#L728-L750

The dpkg-deb command has some parameter that we can pass to indicate the compression? It seems the `-Z` maybe? What will be the correct one?

**Comment by meoso** [comment]
may yes. this may repair 1 of the 2 issues.  (the other being file rights seemingly associated with older NWjs usage)
```
  -Z<type>                         Set the compression type used when building.
                                     Allowed types: gzip, xz, none.
```

unsure the source of the compression method change. im using `ubuntu-latest` elsewhere and it does not introduce this issue by default, but i'm also on very old NWjs/node, so :man_shrugging:

File: src/js/gui.js
Comment (chmelevskij) [inline src/js/gui.js:77]: I like it 🙂

**Review by chmelevskij — APPROVED** [review]

**Review by McGiverGim — APPROVED** [review]
Let's start the versions race again 😜
Let's merge this soon to see if Android works again

File: src/js/gui.js
Comment (McGiverGim) [reply src/js/gui.js:77]: Navigator will be available always, but it does not hurt 😄

**Review by limonspb — APPROVED** [review]

**Comment by haslinghuis** [comment]
@meoso we are on latest now. Did we forget something?

**Comment by meoso** [comment]
> @meoso we are on latest now. Did we forget something?

sorry i had to wait until today (Monday) to return to my Debian box.
```
 $ sudo dpkg -i betaflight-configurator_10.9.0-debug-7145cd6_amd64.deb 
dpkg-deb: error: archive 'betaflight-configurator_10.9.0-debug-7145cd6_amd64.deb' uses unknown compression for member 'control.tar.zst', giving up
```
i cannot test file attributes (chmod rights) until zst resolved.

Portable on Linux works fine.

**Comment by haslinghuis** [comment]
@McGiverGim you have been contributing to` gulp-debian` library before. The project seems to have no activity as I can't find where I would put the `-Z` option. Maybe could be [hard]coded in the `_exec('dpkg-deb --build ...)` function in the library:

https://github.com/stpettersens/gulp-debian/blob/189c808db3d146fcbcaed62182a177bd58aa1b2a/index.js#L194

**Comment by McGiverGim** [comment]
I did some change in the past. If the project is abandoned, there is some alternative?

**Comment by haslinghuis** [comment]
Guess Debian users have to wait for `zst` implementation. Localizing the library is not something I think is a maintainable solution.

**Pull request merged**
